src/main/webapp/location/keyword.tsx

```diff
@@ -0,0 +1,302 @@
+        onClose={() => {
+          setOpen(false)
+          if (keyword && keyword !== input) {
+            setInput(keyword)
+          }
+        }}
+        getOptionLabel={(suggestion: Suggestion) => suggestion.name}
+        options={suggestions}
+        loading={loading}
+        inputValue={input}
+        onChange={(_e, value) => {
+          if (value) {
+            const { id, name } = value
+            setSelectedKeyword({ keywordId: id, keyword: name })
+          }
+        }}
+        renderInput={params => (
+          <TextField
+            {...params}
+            error={error}
+            helperText={open || hasSelection ? undefined : placeholder}
+            label="Keyword"
+            placeholder={`Please enter ${minimumInputLength} or more characters`}
+            fullWidth
+            onChange={e => {
+              setInput(e.target.value)
+            }}
+            InputProps={{
+              ...params.InputProps,
+              endAdornment: (
+                <React.Fragment>
+                  {loading ? (
+                    <CircularProgress color="inherit" size={20} />
+                  ) : null}
+                  {params.InputProps.endAdornment}
+                </React.Fragment>
+              ),
+            }}
+          />
+        )}
+      />
+      {hasSelection ? (
+        <Length
+          label="Buffer"
+          value={{ length: buffer, unit: bufferUnit }}
+          onChange={({ length, unit }) =>
+            onChange(
+              coordinateEditor.polygonPropsToGeo({
+                id,
+                coordinates,
+                buffer: length,
+                bufferUnit: unit,
+                properties,
+              })
+            )
+          }
+        />
+      ) : null}
+    </SpacedLinearContainer>
+  )
+}
+
+const runQuery = async <T extends {}>(
+  promise: Promise<ApolloQueryResult<T>>,
+  setState: (v: QueryState) => void
+): Promise<T> => {
+  setState('loading')
+  let isError = false
+  try {
+    const result = await promise
+    return result.data
+  } catch (e) {
+    setState('error')
+    isError = false
+    throw e
+  } finally {
+    if (!isError) {
+      setState('idle')
+    }
+  }
+}
+
+type suggestionResult = {
+  suggestions: Suggestion[]
+}
+
+type geofeatureResult = {
+  geofeature: geometry.GeometryJSON
+}
+
+const Container: React.SFC<BasicEditorProps> = props => {
+  const client = useApolloClient()
+  const [featureQueryState, setFeatureQueryState] = React.useState<QueryState>(
+    'idle'
+  )
+  const [suggestionQueryState, setSuggestionQueryState] = React.useState<
+    QueryState
+  >('idle')
+  const getGeoFeature = async (id: string) => {
+    return runQuery(
+      client.query<geofeatureResult>({
+        query: FeatureQuery,
+        variables: {
+          id,
+        },
+      }),
+      setFeatureQueryState
+    ).then(r => r.geofeature)
+  }
+  const getSuggestions = async (q: string) => {
+    return runQuery(
```

Comment:
I think you might be able to use `useQuery` here. You can leverage the skip option when you don't have enough information to run the query.

Reply:
Unfortunately useQuery cannot be converted to async functions which is how Keyword is bypassing Apollo in storybook.

Reply:
Why can't useQuery be used with async functions? Could you use `useLazyQuery` instead?

Reply:
I'm going to refactor the async functions out. I'll be able to use `useLazyQuery` after that.

Reply:
I think the keyword component shouldn't be async. It should accept props that can change over time.